src/python/grpcio_tests/tests_aio/unit/interceptor_test.py

```diff
@@ -0,0 +1,340 @@
+
+    async def test_rpc_attributes_available_before_interceptor_ends(self):
+
+        interceptor_lock = asyncio.Event()
+
+        class Interceptor(aio.UnaryUnaryClientInterceptor):
+
+            async def intercept_unary_unary(self, continuation,
+                                            client_call_details, request):
+                call = await continuation(client_call_details, request)
+                await interceptor_lock.wait()
+                return call
+
+        server_target, _ = await start_test_server()  # pylint: disable=unused-variable
+
+        async with aio.insecure_channel(
+                server_target, interceptors=[Interceptor()]) as channel:
+
+            multicallable = channel.unary_unary(
+                '/grpc.testing.TestService/UnaryCall',
+                request_serializer=messages_pb2.SimpleRequest.SerializeToString,
+                response_deserializer=messages_pb2.SimpleResponse.FromString)
+            call = multicallable(messages_pb2.SimpleRequest())
+
+            # Following RPC response attributes can be accessed without
+            # having to wait for the finalitzation of the interceptor code
+            self.assertEqual(await call.code(), grpc.StatusCode.OK)
+            self.assertEqual(await call.details(), '')
+            self.assertEqual(await call.initial_metadata(), ())
+            self.assertEqual(await call.trailing_metadata(), ())
+
+            interceptor_lock.set()
+            await call
+
+    async def test_rpc_response_available_after_interceptor_ends(self):
+
+        interceptor_lock = asyncio.Event()
+
+        class Interceptor(aio.UnaryUnaryClientInterceptor):
+
+            async def intercept_unary_unary(self, continuation,
+                                            client_call_details, request):
+                call = await continuation(client_call_details, request)
+                await interceptor_lock.wait()
+                return call
+
+        server_target, _ = await start_test_server()  # pylint: disable=unused-variable
+
+        async with aio.insecure_channel(
+                server_target, interceptors=[Interceptor()]) as channel:
+
+            multicallable = channel.unary_unary(
+                '/grpc.testing.TestService/UnaryCall',
+                request_serializer=messages_pb2.SimpleRequest.SerializeToString,
+                response_deserializer=messages_pb2.SimpleResponse.FromString)
+            call = multicallable(messages_pb2.SimpleRequest())
+
+            response_read = False
+
+            async def read_response():
+                nonlocal response_read
+                await call
+                response_read = True
+
+            task = asyncio.ensure_future(read_response())
+            await asyncio.sleep(0)
+            self.assertFalse(response_read)
+
+            interceptor_lock.set()
+            await task
+            self.assertTrue(response_read)
+
+    async def test_cancel_before_rpc(self):
```

Comment:
We might need two more test cases that performs before-rpc-cancel and after-rpc-cancel inside of the intercept function.

Reply:
If I'm not missing something, in both test cases, cancellation happens just in the middle of the intercept functions.

Reply:
What I mean is something like this:
```Python
async def intercept_unary_unary(self, continuation, client_call_details, request):
    call = await continuation(client_call_details, request)
    call.cancel()
    await call
    return call
```

```Python
async def intercept_unary_unary(self, continuation, client_call_details, request):
    call = await continuation(client_call_details, request)
    await call
    call.cancel()
    return call
```

Reply:
Good point, I will add them

Reply:
The second one does not make sense to me since the call has been already awaited so it's really cancelable, but I could add it for just double check that nothing gets cancelled